### ISSUE TITLE
Question: using beam decoder in training

### ISSUE THREAD

**Issue by ggdupont** [SeanNaren/deepspeech.pytorch#314]
Maybe a newbie question: I see that the decoder used in training is forced to be GreedyDecoder. Is there a specific reason not to use beam & CTC ?

**Comment by ggdupont** [comment]
Minor progress: I tried to add beam decoder in training. It well surprisingly smooth (dummy copy/pasting from transcribe.py script).

However, it is now stuck on validation at the end of epoch. The BeamCTCDecoder cannot handle call

``decoder.convert_to_strings(split_targets)``

It needs another args as per function description:

``def convert_to_strings(self, out, seq_len):``

seq_len seems to be an array of length, but I can't make it work properly with the information I have in train.py where the ``decoder.convert_to_strings(split_targets)`` call is made. Any clue?

**Comment by miguelvr** [comment]
Legitimate question: Is the BeamCTCDecoder tractable? i.e. can you backpropagate through the beam search? I'm not sure...

**Comment by ryanleary** [comment]
We use the greedy decoder because the decoded results are not taken into consideration as part of the training process. What is being optimized is the CTC loss. Doing the decoding to get WER/LER values per epoch is really just a sanity check for the user/practitioner. The loss and LER (and to a lesser extent WER) should be roughly correlated. 

To @miguelvr I'm not entirely sure without spending some serious time thinking how one might improve on the loss function to take into account the decoded results.
